pkg/provider/bitbucketserver/bitbucketserver.go

```diff
@@ -229,3 +239,68 @@ func (v *Provider) GetConfig() *info.ProviderConfig {
 		TaskStatusTMPL: taskStatusTemplate,
 	}
 }
+
+// Detect processes event and detect if it is a bitbucket server event, whether to process or reject it
+// returns (if is a bitbucket server event, whether to process or reject, error if any occurred)
+func (v *Provider) Detect(reqHeader *http.Header, payload string, logger *zap.SugaredLogger) (bool, bool, *zap.SugaredLogger, error) {
+	isBitServer := false
+	event := reqHeader.Get("X-Event-Key")
+	if event == "" {
+		return false, false, logger, nil
+	}
+
+	eventPayload, err := parsePayloadType(event)
+	if err != nil {
+		return false, false, logger, err
+	}
+	if eventPayload == nil {
+		return false, false, logger, nil
+	}
+	// it is a Bitbucket server event
+	isBitServer = true
+
+	setLoggerAndProceed := func() (bool, bool, *zap.SugaredLogger, error) {
+		logger = logger.With("provider", "bitbucket-server", "event", reqHeader.Get("X-Request-Id"))
+		return isBitServer, true, logger, nil
+	}
+
+	if err := json.Unmarshal([]byte(payload), &eventPayload); err != nil {
+		return isBitServer, false, logger, err
+	}
+
+	switch e := eventPayload.(type) {
+	case *types.PullRequestEvent:
+		if valid(event, []string{"pr:from_ref_updated", "pr:opened"}) {
+			return setLoggerAndProceed()
+		}
+		if valid(event, []string{"pr:comment:added", "pr:comment:edited"}) {
+			if matches, _ := regexp.MatchString(retestRegex, e.Comment.Text); matches {
+				return setLoggerAndProceed()
+			}
+			if matches, _ := regexp.MatchString(oktotestRegex, e.Comment.Text); matches {
+				return setLoggerAndProceed()
+			}
+		}
+		return isBitServer, false, logger, nil
+
+	case *types.PushRequestEvent:
+		if valid(event, []string{"repo:refs_changed"}) {
+			if e.Changes != nil {
+				return setLoggerAndProceed()
+			}
+		}
+		return isBitServer, false, logger, nil
+
+	default:
+		return isBitServer, false, logger, fmt.Errorf("event %s is not supported", event)
+	}
+}
+
+func valid(value string, validValues []string) bool {
```

Comment:
same as before, no need to copy and paste if we can common it,

Reply:
same as above
let me handle it